### PR TITLE
Dbc file fixes

### PR DESCRIPTION
**Pull request by juleq**
I followed your advice, the commits are only two now, but the diffs are still several...

### CODE REVIEW

**Comment by eerimoq** [comment]
Nice with fewer commits.

I think you have to manually fix the unwanted diffs.

**Comment by juleq** [comment]
I would do that by merging with upstream/master, right?

**Comment by eerimoq** [comment]
`git checkout upstream/master <file to replace with upstream/master version>`, then commit and push.

**Comment by juleq** [comment]
Did it manually. Should be good now.

File: tests/files/foobar.dbc [hunk 1]
```diff
@@ -1,7 +1,7 @@
-VERSION "2.0"
+VERSION ""
```
[the remark this answers is not in the document]
Comment (juleq) [reply tests/files/foobar.dbc]: CANdb++ removed it so I trusted it :)

File: tests/files/foobar.dbc [hunk 2]
```diff
@@ -33,56 +33,66 @@ NS_ :
...
- SG_ Fam : 12|12@1- (1.0,0.0) [0|1] "" Vector__XXX, BAR
+ SG_ Fum : 0|12@1- (1,0) [0|1] ""  BAR
+ SG_ Fam : 12|12@1- (1,0) [0|1] ""  BAR
 
 BO_ 2147558194 Bar: 4 FOO
- SG_ Binary32 : 0|32@1- (1,0) [0|0] "" Vector__XXX, FUM
+ SG_ Binary32 : 0|32@1- (1,0) [0|0] ""  FUM
 
 BO_ 2147558195 CanFd: 64 FOO
- SG_ Foo : 0|512@1- (1,0) [0|0] "" Vector__XXX, FUM
```
[the remark this answers is not in the document]
Comment (juleq) [reply tests/files/foobar.dbc]: Yes. There are only types up to 64 bits. Choosing the length to be greater will get you an error.

**Comment by eerimoq** [comment]
After removing [this](https://github.com/eerimoq/cantools/blob/master/tests/test_database.py#L91) a non-empty DBC version is not tested anymore. Does CANdb++ remove all versions, everywhere? Why?

**Comment by juleq** [comment]
I have done the following to find out what version does:
- create a new dbc - no options with respect to version
- save as - no options with respect to version
- export - no options with respect to version
- scanned through help - no mentions of the version tag
- web search - nothing

And if you put one in, is does remove it on save. We can put it back in for the tests, but I think it might be abandoned. At least I cant´t find a either purpose or willingness to retain that info. Strange.

**Comment by eerimoq** [comment]
Let's put it back in. It might be added in later versions of CANdb++, or similar tools if there are such.

**Comment by juleq** [comment]
By the way: The fixed do only apply when manipulating the objects: I stumbled over this myself when I created new messages in code without setting a dbc object and then dumping them out.

File: tests/test_database.py [hunk 3]
```diff
@@ -1837,6 +1847,24 @@ def test_lookups(self):
...
+        node = NodeMock("FakeNode", "Comment")
+        nodes.append(node)
+        db = cantools.db.Database(nodes=nodes)
+        signals = []
+        signals.append(cantools.db.Signal("C", 0, 8))
+        db.messages.append(cantools.db.Message(0x20, "D", 8, signals))       
+
+        # Test if dump executes without raising
+        db.as_dbc_string()
+
```
Comment (eerimoq) [inline tests/test_database.py]: I suggest first creating an empty database and then add the node and message, or is there a reason to give the node as argument to the database creation, and then add the message after the database has been created?

Add `from .node import Node` in `cantools/db/__init__.py` to use it as below.

``` python
def test_missing_dbc_specifics(self):
    db = cantools.db.Database()

    node = cantools.db.Node('FakeNode', 'Comment')
    db.nodes.append(node)

    signals = [cantools.db.Signal('C', 0, 8)]
    message = cantools.db.Message(0x20, 'D', 8, signals)
    db.messages.append(message)       

    # Test that dump executes without raising.
    db.as_dbc_string()
```

Comment (juleq) [reply tests/test_database.py]: Yes, that was what I was missing...

**Comment by juleq** [comment]
The change is just a fix for an edge case (programmatically creating the database and dump it as dbc). I get your point though and would propose to branch out for non-related stuff in the future. Ok?

**Comment by eerimoq** [comment]
Many PR:s with one feature or fix is preferred over one PR with lots of features and fixes. If branching help you achieve that, go for it =)

Is this PR ready to be merged now?

**Comment by juleq** [comment]
Yes, it is ready to be merged. I will test as soon as you have that branch available.

**Comment by eerimoq** [comment]
The DBC line endings looks correct already. I'll not make any changes.

**Comment by juleq** [comment]
Where should we put to do items? E.g. I saw that VAL_TABLE_ dumping is not implemented yet. Should I create an issue to keep track of that or do you want a todo.md file or something?

**Comment by eerimoq** [comment]
Create an issue for it and label it enhancement or something similar.
